ports/mimxrt: Some misc fixes to sdio, sdram and Makefile.

iabdalkader opened the pull request:
- mimxrt/mbedtls: Enable certificate validity time validation.
- ports/mimxrt: Enable flow control for BT HCI Uart.
- ports/mimxrt: Remove SDCARD Makefile config option.
- ports/mimxrt/Makefile: Enable the FSL USDHC for supported MCU series.
- ports/mimxrt: Allow boards to override the default SDRAM config.
- ports/mimxrt/sdio.c: Add support for the 117x series.

FWIW, the existing SDRAM timing is off for both 117x series and other series, but I kept them the same.

iabdalkader committed:
mimxrt/sdio: Add support for the 117x series.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
mimxrt/mimxrt_sdram: Allow boards to override the default SDRAM config.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
mimxrt/Makefile: Enable the FSL USDHC for supported MCU series.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader commented on ports/mimxrt/Makefile:
```diff
@@ -152,7 +152,7 @@ ifeq ($(MICROPY_HW_SDRAM_AVAIL),1)
 endif
 
-ifeq ($(MICROPY_PY_MACHINE_SDCARD),1)
+ifeq (1, $(filter 1, $(MICROPY_PY_MACHINE_SDCARD) $(MICROPY_PY_NETWORK_CYW43)))
```
Yes, done.

robert-hh replied:
The respective MICROPY_PY_MACHINE_SDCARD settings in the mpconfigboard.h files are then obsolete.

iabdalkader replied:
Yes, it should be moved to boardconfig files.

iabdalkader replied:
Maybe also renamed to `MICROPY_HW_ENABLE_SDCARD` to be consistent with other ports.

robert-hh replied:
> Yes, it should be moved to boardconfig files

It is ATM in the boardconfig files. But this PR moves it to mpconfigport.h.

iabdalkader replied:
Yes it's currently in `mpconfigboard.mk`s, I was saying it should be in `mpconfigboard.h`, and renamed to `MICROPY_HW_ENABLE_SDCARD` to be consistent with boards.

iabdalkader replied:
But it's not needed as a Makefile option anymore because the USDHC is compiled per series, so if moved to mpconfigboard.h files, it will work, see the last commit, if you have a different idea we can drop it.

robert-hh replied:
> because the USDHC is compiled per series,

I've seen that. And because of that we can remove the individual setting in each board file set and just use one setting in Makefile or mpconfigport.h.

iabdalkader replied:
Oh okay, I see what you mean, so it will be enabled by default unless defined in `mpconfigboard.h` ?

robert-hh replied:
About. I had expected to set  `MICROPY_PY_MACHINE_SDCARD` after the line 
`ifeq ($(MCU_SERIES),$(filter $(MCU_SERIES), MIMXRT1021 MIMXRT1052 MIMXRT1062 MIMXRT1064 MIMXRT1176))`
in Makefile. But the actual state is fine. Didn't you want to rename it to `MICROPY_HW_ENABLE_SDCARD`?

iabdalkader replied:
> Didn't you want to rename it to `MICROPY_HW_ENABLE_SDCARD`?

I did, but I think it's now more consistent with the other `MICROPY_PY_MACHINE_xxx` options in `mpconfigport.h`. There are 3 ports that have `machine_sdcard.c`, when it becomes standard, I think the correct option to enable it would be  `MICROPY_PY_MACHINE_SDCARD`, the `MICROPY_HW_ENABLE_SDCARD` should be for just enabling `sdcard.c` (for example for C-level usage, sdio.c etc..).

iabdalkader committed:
mimxrt: Remove SDCARD Makefile config option.

This is option is no longer needed as a Makefile option as the USDHC driver
is enabled for all supported series.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

kwagyeman commented on ports/mimxrt/mimxrt_sdram.c:
```diff
@@ -209,62 +250,38 @@ void mimxrt_sdram_init(void) {
+        .tPrecharge2Act_Ns = MICROPY_HW_SDRAM_TIMING_TRP,
+        .tAct2ReadWrite_Ns = MICROPY_HW_SDRAM_TIMING_TRCD,
+        .tRefreshRecovery_Ns = MICROPY_HW_SDRAM_TIMING_TXSR,
+        .tSelfRefRecovery_Ns = MICROPY_HW_SDRAM_TIMING_TXSR,
```
Not clear what tRefreshRecovery_Ns and tSelfRefRecovery_Ns are. Seems like they are TXSR. Not sure why there are two. Previous value of 60 + 67 seems wrong. Not value that high on the datasheet.

So, I think the above changes are correct.

kwagyeman replied:
Looking at other datasources... this is the refresh to active time. So, seems correct. Same value I guess.

kwagyeman commented on ports/mimxrt/mimxrt_sdram.c:
```diff
@@ -42,10 +42,51 @@ extern uint8_t __sdram_start;
+#define MICROPY_HW_SDRAM_RBURST_LENGTH      (1)
+#define MICROPY_HW_SDRAM_DELAY_CHAIN        (2)
+#else
+#define MICROPY_HW_SDRAM_TIMING_TRC         (60)
```
Datasheet reference: https://www.mouser.com/datasheet/2/198/42-45S83200G-16160G-258274.pdf

kwagyeman commented on ports/mimxrt/mimxrt_sdram.c:
```diff
@@ -42,10 +42,51 @@ extern uint8_t __sdram_start;
+#define MICROPY_HW_SDRAM_TIMING_TRDD        (60)
+#define MICROPY_HW_SDRAM_TIMING_TXSR        (67)
+#define MICROPY_HW_SDRAM_TIMING_TRAS        (42)
+#define MICROPY_HW_SDRAM_TIMING_TREF        (64 * 1000000 / 8192) // 64ms/8192
```
Err, use the right column. Same numbers.

kwagyeman commented on ports/mimxrt/mimxrt_sdram.c:
```diff
@@ -42,10 +42,51 @@ extern uint8_t __sdram_start;
+#define MICROPY_HW_SDRAM_TIMING_TREF        (64 * 1000000 / 8192) // 64ms/8192
+
+#define MICROPY_HW_SDRAM_CAS_LATENCY        (kSEMC_LatencyThree)
+#define MICROPY_HW_SDRAM_MEM_BUS_WIDTH      (kSEMC_PortSize16Bit)
```
Correct

kwagyeman commented on ports/mimxrt/mimxrt_sdram.c:
```diff
@@ -42,10 +42,51 @@ extern uint8_t __sdram_start;
+
+#define MICROPY_HW_SDRAM_CAS_LATENCY        (kSEMC_LatencyThree)
+#define MICROPY_HW_SDRAM_MEM_BUS_WIDTH      (kSEMC_PortSize16Bit)
+#define MICROPY_HW_SDRAM_COLUMN_BITS_NUM    (kSEMC_SdramColunm_9bit)
```
Correct

iabdalkader commented:
@kwagyeman Thanks for double checking the numbers, but you are looking at a different part's datasheet, I kept the original timings for 106x/117x to avoid breaking anything, there's no need to review them.

robert-hh commented:
When the code for RAM was initially written, both @alphaFred and me cross-checked the timings for the RAM. Which does not exclude that there may be still wrong numbers. But at least it worked until now and showed not problems.

iabdalkader commented:
@robert-hh I don't think any set of numbers will work for all boards, so for now it's more important to just make it possible to override the configuration. However, the most obvious issue with the numbers are `tWR` and `tRRD` for 117x, both use 2ns which I think is a mistake, in the datasheet the numbers for those are 2 but the units are in tCK so at 160MHz they should be 12ns (which happens to be the same for a number of other SDRAM parts).

iabdalkader committed:
mimxrt/mpbthciport: Enable flow control for BT HCI UART.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
mimxrt/mbedtls: Enable certificate validity time validation.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

dpgeorge commented:
Merged.
